In [1]:
!pip install transformers

In [10]:
!sudo pip install gluonnlp

Password:
sudo: a password is required


In [2]:
import torch


In [14]:
from transformers import BertModel, BertTokenizer

tokenizer = BertTokenizer.from_pretrained('bert-base-cased')
model = BertModel.from_pretrained('bert-base-cased')

Downloading:   0%|          | 0.00/213k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/436M [00:00<?, ?B/s]

In [16]:
tokenizer

PreTrainedTokenizer(name_or_path='bert-base-cased', vocab_size=28996, model_max_len=512, is_fast=False, padding_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'})

In [17]:
tokenizer.encode("I am a student")

[101, 146, 1821, 170, 2377, 102]

In [19]:
tokenizer.decode(tokenizer.encode("I am a student"))

'[CLS] I am a student [SEP]'

In [31]:
inputs = tokenizer.encode_plus("I am a student", return_tensors='pt')
inputs

{'input_ids': tensor([[ 101,  146, 1821,  170, 2377,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1]])}

In [35]:
out = model(input_ids=inputs['input_ids'],
            token_type_ids=inputs['token_type_ids'],
            attention_mask=inputs['attention_mask'])
# 같은 표현
# out = model(**inputs) 

In [38]:
# 문장을 임베딩한 값
# (bacth, dim)
out.pooler_output.shape

torch.Size([1, 768])

In [40]:
# 각 토큰을 임베딩한 값 
# (batch, token, dim)
out.last_hidden_state.shape

torch.Size([1, 6, 768])


## 영화 감성 분석을 할 경우
1. 문장과 긍/부정으로 이루어진 clasffication 문제 
2. 모든 문장을 버트 모델에 태움
3. 마지막 layer의 pooler output 가져옴
4. 768 -> 2 fully connected layer를 생성
5. 해당 레이어 학습
6. 감성 분석 가능
